### PR TITLE
Add new sync config: l2.enginekind

### PR DESCRIPTION
**Pull request by mininny**
tl;dr: We need to enable EL sync to start from the middle of the chain for archive node specialized clients—e.g., op-erigon. I’d like to propose a new syncmode `force-execution-layer` in op-node that would skip the initial validation where no finalized block in the node should exist.

## Summary
Currently, op-node only uses execution-layer syncing if there is no finalized block in the execution client or if that finalized block is the bedrock transition block ([ref](https://github.com/ethereum-optimism/optimism/blob/7cbda018196b58a71e2c0b4bc9e31a289235074e/op-node/rollup/derive/engine_controller.go#L331-L345)). When the execution client has a finalized block, it is able to build further blocks from that block, so regular consensus-layer syncing is feasible. This approach makes sense for op-geth, which performs initial snap sync from scratch, where the finalized block is created only once the sync has completed. After the initial snap sync has completed, geth transitions to full block-by-block sync, which aligns with the op-node's behavior.

However, alternative archival execution clients like op-erigon retrieve every block from the genesis, taking longer time to sync. In order to reduce the sync time, new erigon nodes uses erigon [snapshot](https://github.com/ledgerwatch/erigon/tree/main/cmd/downloader) (torrent of the block data), which is downloaded and used to start syncing from. In addition, archive of the datadir are also often provided to download a more recent snapshot of the whole db.

Because using above methods to quick-start the syncing will end up in db with weeks of unsynced data, it would still beneficial to allow execution-layer from an arbitrary block to the latest head using p2p execution-layer syncing. This way, we can use various snapshots to download the majority of the chain data, and use p2p syncing to quickly catch up to the chain's head.

## Proposed Solution
Since we need to preserve the behavior of op-node’s sync mode transitioning from `execution-layer` to `consensus-layer` for op-geth, I propose introducing a new sync mode, such as `force-execution-layer`. 

This flag should skip op-node’s initial check for a finalized block in the client and always start syncing with execution-layer mode until the execution client returns with `VALID`. Once the execution client has synced to the latest known head, it can then revert to consensus layer syncing. Unlike `execution-layer`, this won't switch to `consensus-layer` automatically. 

## Implementation
- Add a new option for `sync.SyncMode`: `ForceELSync`
    - This new mode should act identical to `ELSync` mode in most places
- In `engine_controller.InsertUnsafePayload`, use `ForceELSync` mode to always to set `e.syncStatus = syncStatusStartedEL`
    - We could also consider adding a brand new flag to be used alongside `syncmode=execution-layer`, like: `strict-syncmode=true`
    - This adds the complexity of user and flag setting in the codebase, so we can just add a new alias to `ELSync` as `ForceELSync` to implicit contain this meaning.
- For the sync between op-node and the execution client, most of the behavior remains unchanged:
    - op-node provides the tip of the chain to sync towards
        - Execution client will perform internal p2p sync to sync between the tip of the chain and the latest finalized head in execution client.
        - the payload is `headBlockHash: headBlockHash`, `safeBlockHash: 0x00`, `finalizedBlockHash: 0x00`
        - As the node is syncing, it will return SYNCING status to FCU and newPayload
    - One the execution client completes syncing, it returns VALID to op-node
        - This will update the engine’s head/safe/finalized block hash to the synced target.
    - This will trigger the conversion from ELSync to CLSync, and op-node will derive each block.

## Pros
- This can be helpful for archival nodes that have gone offline for a while. If we resume syncing with p2p, it would catch up to the head much faster than building each block with the op-node.
- op-reth also uses erigon's staged sync architecture. Introducing this would help the syncing of op-reth as well.

## Concerns
- If the node's p2p network is stale, the node can be stuck in `force-execution-layer` mode
    - This is still the case in new nodes in `execution-layer` syncmode if it can't connect to any peers.
- Using `force-execution-layer` requires the user to be knowledgeable of the how the execution client syncs, and whether it's able to initiate p2p sync in the middle of chain.
    - If you run op-geth that has been synced before with `force-execution-layer` it will not be able to sync as it  already has transitioned to full sync

## Alternative Considered
- Format of the flag
    - The name `force-execution-layer` could be somewhat confusing to the user without this much background information.
    - Instead, we could introduce something like `l2.enginekind` which is like the `l1.rpckind` which informs node what kind of client it is communicating with.
    - It should also have presets like `geth`, `erigon`, `reth` , with `standard` being likely geth.
    - When the `l2.enginekind` is `erigon` |`reth` , we could skip the initial finalized block check, which is same as the above implementation shown.

## Related Notes
- Based on op-node code, there is `protocol-quest/91` which seems to be tracking this

---

If there are any caveats or complications in different syncmodes that this PR fails to address, please let me know! I'd be glad to take this back and reconsider it.

### CODE REVIEW

**Comment by axelKingsley** [comment]
I really like the ideas proposed in the description here, thank you for putting all that context together!

This issue highlights a fundamental issue around EL management from OP-Node: aspects of the Engine which are not based in ethereum-spec may behave differently on different clients. Synchronization is the largest non-conforming behavior I can think of, but I bet there are more (metrics, for example).

So, the suggestion of introducing `l2.engineKind` is really great, and imo a desired thing to introduce to keep our synchronization logic clean if/as more clients get added to the OP Stack ecosystem. I prefer this solution over `force-execution-layer`, because (as you rightly mention) it obfuscates that these modes are actually just mapping to different clients. These sync-mode labels would start to lose meaning as we incorporate more clients with differing features, and they would de-facto become client-specific.

Regarding the code in this PR -- It's not clear to me if this is code-complete. It seems like it introduces the new flag, and then unifies the flag by calling `IsSyncMode()`, but I don't see any difference in behavior depending on which flag is introduced. And, I believe you've spoken to @sebastianst about testing, which I agree, would be good to see to confirm that this has the intended effect on Erigon and does not affect Geth.

File: op-node/rollup/derive/engine_controller.go [hunk 1]
```diff
@@ -331,7 +331,7 @@ func (e *EngineController) InsertUnsafePayload(ctx context.Context, envelope *et
 	if e.syncStatus == syncStatusWillStartEL {
 		b, err := e.engine.L2BlockRefByLabel(ctx, eth.Finalized)
 		rollupGenesisIsFinalized := b.Hash == e.rollupCfg.Genesis.L2.Hash
-		if errors.Is(err, ethereum.NotFound) || rollupGenesisIsFinalized {
+		if errors.Is(err, ethereum.NotFound) || rollupGenesisIsFinalized || e.syncMode == sync.ForceELSync {
```
Comment (sebastianst) [inline op-node/rollup/derive/engine_controller.go]: @axelKingsley this is the crucial diff added in this PR.

**Comment by sebastianst** [comment]
I agree with @axelKingsley assessment. If there's undefined behavior around engines, then an `engine-kind` flag could be semantically clearer than the change introduced in this PR. Thanks for working through these changes with us 🙏🏻

**Comment by mininny** [comment]
@axelKingsley @sebastianst Thank you guys for such thorough input! 😄 I agree that `l2.enginekind` flag looks much clearer. 

I modified the code to introduce a new `l2.enginekind` and use it to configure the sync behavior in the op-node. 

You can also find relevant tests in `sync_test.go`. These tests were using op-geth implementation, so actually forcing the engine to perform a EL Sync with a finalized block was not possible (as geth transitions to full sync). So, I just added validations that checks op-node's logs regarding transition to CL Sync or starting EL Sync. 

I'm still new with the codebase so please let me know if you think some code are better placed elsewhere! 🙏

File: op-program/client/driver/driver.go [hunk 2]
```diff
@@ -101,7 +101,7 @@ type Driver struct {
 }
 
 func NewDriver(logger log.Logger, cfg *rollup.Config, l1Source derive.L1Fetcher, l1BlobsSource derive.L1BlobsFetcher, l2Source L2Source, targetBlockNum uint64) *Driver {
-	engine := engine.NewEngineController(l2Source, logger, metrics.NoopMetrics, cfg, sync.CLSync)
+	engine := engine.NewEngineController(l2Source, logger, metrics.NoopMetrics, cfg, &sync.Config{SyncMode: sync.CLSync})
 	attributesHandler := attributes.NewAttributesHandler(logger, cfg, engine, l2Source)
 	syncCfg := &sync.Config{SyncMode: sync.CLSync}
```
Comment (axelKingsley) [inline op-program/client/driver/driver.go]: I notice that these sync configs never set their L2 Engine Kind, should they have it set?

(also, could we move `sysCfg :=` to a higher line so that the engine create can use it too?)

File: op-node/rollup/sync/config.go [hunk 3]
```diff
@@ -70,4 +111,7 @@ type Config struct {
 	// Note: We probably need to detect the condition that snap sync has not complete when we do a restart prior to running sync-start if we are doing
 	// snap sync with a genesis finalization data.
 	SkipSyncStartCheck bool `json:"skip_sync_start_check"`
+
+	// Engine Client's kind. This helps us determine some of the behavior of the op-node.
+	L2EngineClientKind EngineClientKind `json:"l2_engine_client_kind"`
```
Comment (axelKingsley) [inline op-node/rollup/sync/config.go]: I would have expected this field to exist at the parent-scope of the config (like at the top level, or if we have some Engine-Control section of config), and then to be passed down into Sync Configs via `NewSyncConfig`, or potentially even designated by `NewEngineController`, as the controller would have the most authority over the engine-kind.

This flag is currently only needed by Sync, so it's ok to be owned there, but if in the future we see more differences in the ways we need to treat engines, we would want to pass this engine-kind down to all the configurations from a central point.

File: op-e2e/actions/l2_verifier.go [hunk 4]
```diff
@@ -31,10 +31,7 @@ import (
 type L2Verifier struct {
 	log log.Logger
 
-	eng interface {
-		engine.Engine
-		L2BlockRefByNumber(ctx context.Context, num uint64) (eth.L2BlockRef, error)
-	}
```
Comment (axelKingsley) [inline op-e2e/actions/l2_verifier.go:43]: This seems safe, changing `eng` to an `L2API` just means it has to support more functions,
```
InfoByHash(ctx context.Context, hash common.Hash) (eth.BlockInfo, error)
GetProof(ctx context.Context, address common.Address, storage []common.Hash, blockTag string) (*eth.AccountResult, error)
OutputV0AtBlock(ctx context.Context, blockHash common.Hash) (*eth.OutputV0, error)
```
Which it probably does, given that it compiles :). Just wanted to understand the motivation of the change, are there related changes in the rest of this PR?

File: op-node/rollup/engine/engine_controller.go [hunk 5]
```diff
@@ -75,19 +76,25 @@ type EngineController struct {
...
 	syncStatus := syncStatusCL
-	if syncMode == sync.ELSync {
+	if syncCfg.SyncMode == sync.ELSync {
 		syncStatus = syncStatusWillStartEL
 	}
 
+	engineKind := syncCfg.L2EngineClientKind
+	if !sync.ValidEngineClientKind(syncCfg.L2EngineClientKind) {
+		engineKind = sync.EngineClientGeth
+	}
```
Comment (axelKingsley) [inline op-node/rollup/engine/engine_controller.go]: (related to the comment I left on `config.go` regarding where this flag should live)

It reads a little strangely for `NewEngineController` to only learn about the engine-kind through the `syncCfg`

**Review by axelKingsley — COMMENTED** [review]
Looking good! I left some comments regarding where this flag should live in terms of config scope, as it feels like a value that would be more appropriate being passed-in to the syncConfig from some higher level. Let me know what you think, I could be way off-base.

File: op-node/rollup/engine/engine_controller.go [hunk 6]
```diff
@@ -376,7 +383,7 @@ func (e *EngineController) InsertUnsafePayload(ctx context.Context, envelope *et
 	if e.syncStatus == syncStatusWillStartEL {
 		b, err := e.engine.L2BlockRefByLabel(ctx, eth.Finalized)
 		rollupGenesisIsFinalized := b.Hash == e.rollupCfg.Genesis.L2.Hash
-		if errors.Is(err, ethereum.NotFound) || rollupGenesisIsFinalized {
+		if errors.Is(err, ethereum.NotFound) || rollupGenesisIsFinalized || e.engineKind != sync.EngineClientGeth {
```
Comment (sebastianst) [inline op-node/rollup/engine/engine_controller.go]: Wondering if we really want to use _exclusion_ here as opposed to explicit _inclusion_? If we ever add another engine kind, this would silently apply to that new engine kind using exclusion.

File: op-e2e/actions/l2_verifier.go (hunk 4 above)
Comment (mininny) [reply op-e2e/actions/l2_verifier.go:43]: I modified this so that I could use the existing eng from the L2Verifier for the NewL2Verifier creation. This is used in tests where a new node is "relaunched" while the execution engine stays the same. :)

File: op-node/rollup/sync/config.go (hunk 3 above)
Comment (mininny) [reply op-node/rollup/sync/config.go]: I agree with your comments on the location of this new flag. I think it makes sense to treat it independently from the Sync Configs; I modified it so it's now parsed in `node.Config` and passed to the `EngineController` which will use the value directly. https://github.com/ethereum-optimism/optimism/pull/10767/commits/8c84cf3bfee500d61de592cf49f579cf502647f7

This did add a lot of boilerplate code everywhere, so I'd love to know whether you think this much is fine :)

File: op-program/client/driver/driver.go (hunk 2 above)
Comment (mininny) [reply op-program/client/driver/driver.go]: Mostly, I think it will only be relevant when it's set to something other than Geth from the op-node config. 
But I modified the code so that they have default L2 Engine Kind of `EngineClientGeth`. And I moved the `sysCfg :=` to above!

File: op-node/rollup/engine/engine_kind.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+import "fmt"
+
+// EngineClientKind identifies the engine client's kind, used to control the behavior of optimism in different engine clients.
+type EngineClientKind string
+
+const (
+	EngineClientGeth   EngineClientKind = "geth"
+	EngineClientReth   EngineClientKind = "reth"
+	EngineClientErigon EngineClientKind = "erigon"
```
Comment (sebastianst) [inline op-node/rollup/engine/engine_kind.go]: I propose to shorten all occurrences of `EngineClient` to `Engine` everywhere, e.g.
```suggestion
type EngineKind string

const (
	EngineGeth   EngineKind = "geth"
	EngineReth   EngineKind = "reth"
	EngineErigon EngineKind = "erigon"
```

I don't think that omitting the `Client` leads to any ambiguity.

File: op-node/rollup/engine/engine_controller.go [hunk 8]
```diff
@@ -75,9 +76,9 @@ type EngineController struct {
 	safeAttrs    *derive.AttributesWithParent
 }
 
-func NewEngineController(engine ExecEngine, log log.Logger, metrics derive.Metrics, rollupCfg *rollup.Config, syncMode sync.Mode) *EngineController {
+func NewEngineController(engine ExecEngine, log log.Logger, metrics derive.Metrics, rollupCfg *rollup.Config, syncCfg *sync.Config, engineKind EngineClientKind) *EngineController {
```
Comment (sebastianst) [inline op-node/rollup/engine/engine_controller.go]: @axelKingsley wondering if adding the engineKind to the syncConfig is actually a clean place for it to live? Currently, the receives of sync configs are also the ones interested in the engine kind.

File: op-e2e/actions/sync_test.go [hunk 9]
```diff
@@ -830,6 +832,264 @@ func TestELSyncTransitionstoCL(gt *testing.T) {
 	require.Equal(t, uint64(23), id.Number)
 }
 
+func TestELSyncTransitionsToCLSyncAfterNodeRestart(gt *testing.T) {
```
Comment (sebastianst) [inline op-e2e/actions/sync_test.go:790]: If the new tests are mostly copied over from another test, it would be nice to actually only define one function that contains most testing logic and then add some parameterization, to achieve deduplication.

File: op-node/rollup/sync/config.go (hunk 3 above)
Comment (axelKingsley) [reply op-node/rollup/sync/config.go]: I'm okay with the additional wiring that's been added -- it makes sense that a new top-level value would need to be added to signatures.

In the future, if we ever need to create an entire `engine.Config`, it'll be easy to do so now too :).

File: op-node/rollup/engine/engine_kind.go (hunk 7 above)
Comment (sebastianst) [reply op-node/rollup/engine/engine_kind.go]: Or seeing that they are in package `engine`, maybe even drop the `Engine`. That would be idiomatic in Go, so they can be used as `engine.Kind`, `engine.Geth` etc.

File: op-node/rollup/engine/engine_controller.go (hunk 6 above)
Comment (axelKingsley) [reply op-node/rollup/engine/engine_controller.go]: Good point, though when introducing a new Engine, without knowing details it's not possible to know if default-true or default-false is a better option.

Not something we need to have in this PR at all, but down the line we could do something like `|| e.engineKind.SyncBehaviorDoesTheThing()`, and then all the various engine kinds can define their preferred behavior somewhere else.

**Review by axelKingsley — APPROVED** [review]

File: op-node/rollup/engine/engine_controller.go (hunk 6 above)
Comment (mininny) [reply op-node/rollup/engine/engine_controller.go]: > Not something we need to have in this PR at all, but down the line we could do something like || e.engineKind.SyncBehaviorDoesTheThing(), and then all the various engine kinds can define their preferred behavior somewhere else.

I like this idea since we can organize how each behavior is defined by the engine kind. Trying to think of a good naming; what do you think about something like `engineKind.SupportsPostFinalizationELSync` or `engineKind.CanContinueELSync`?

File: op-node/rollup/engine/engine_kind.go (hunk 7 above)
Comment (mininny) [reply op-node/rollup/engine/engine_kind.go]: I agree that it looks much better! I'll modify it to Engine.Kind / Engine.Geth.

File: op-node/rollup/engine/engine_controller.go (hunk 6 above)
Comment (axelKingsley) [reply op-node/rollup/engine/engine_controller.go]: between these two, `SupportsPostFinalizationELSync` gives me a greater understanding

**Comment by mininny** [comment]
@axelKingsley @sebastianst Thanks for the review guys! I addressed all of your comments, and I think it can be merged if you guys think it's fine 😃

File: op-program/client/driver/driver.go [hunk 10]
```diff
@@ -101,9 +101,9 @@ type Driver struct {
 }
 
 func NewDriver(logger log.Logger, cfg *rollup.Config, l1Source derive.L1Fetcher, l1BlobsSource derive.L1BlobsFetcher, l2Source L2Source, targetBlockNum uint64) *Driver {
-	engine := engine.NewEngineController(l2Source, logger, metrics.NoopMetrics, cfg, sync.CLSync)
-	attributesHandler := attributes.NewAttributesHandler(logger, cfg, engine, l2Source)
 	syncCfg := &sync.Config{SyncMode: sync.CLSync}
+	engine := engine.NewEngineController(l2Source, logger, metrics.NoopMetrics, cfg, syncCfg, engine.Geth)
```
Comment (sebastianst) [inline op-program/client/driver/driver.go]: I suggest to name the variable different from `engine` because it now shadows the `engine` package.

File: op-e2e/actions/sync_test.go [hunk 11]
```diff
@@ -807,27 +782,146 @@ func TestELSyncTransitionstoCL(gt *testing.T) {
...
 
 	// Create 1 more block & batch submit everything
+	BatchSubmitBlock(t, miner, sequencer, verifier, batcher, dp, 12)
+
+	// Verify that the batch submitted blocks are there now
+	VerifyBlock(t, sequencer.eng, 23, eth.Safe)
+	VerifyBlock(t, verifier.eng, 23, eth.Safe)
+}
+
+func TestELSyncTransitionsToCLSyncAfterNodeRestart(gt *testing.T) {
```
Comment (sebastianst) [inline op-e2e/actions/sync_test.go:790]: This test and `TestForcedELSyncCLAfterNodeRestart` still seem to share a lot of common code. Is there a way to deduplicate?

**Review by sebastianst — COMMENTED** [review]
LGTM, just some minor comments.

File: op-node/rollup/driver/driver.go [hunk 12]
```diff
@@ -172,13 +172,14 @@ func NewDriver(
 	syncCfg *sync.Config,
 	sequencerConductor conductor.SequencerConductor,
 	plasma PlasmaIface,
+	engineClientKind engine.Kind,
```
Comment (protolambda) [inline op-node/rollup/driver/driver.go]: Would it not be better to hide the "engine kind" from the main logic, and instead translate the engine-kind CLI flag into config values in the `sync.Config`? A `SupportsPostFinalizationELSync` config attribute in the `sync.Config` makes more sense than reading potentially conflicting sync-configuration from two different objects.

**Review by protolambda — COMMENTED** [review]
Functionality is ok, but I would prefer we approach the internal configuration part differently, to not leak the engine-kind detail, and keep consistency with the existing sync-config. The external facing config flag of the engine-kind is nice to have though, since that is easier to understand by end-users than individual specialized config-attribute flags.

File: op-program/client/driver/driver.go (hunk 10 above)
Comment (mininny) [reply op-program/client/driver/driver.go]: Thank you! fixed :) https://github.com/ethereum-optimism/optimism/pull/10767/commits/d2573168729a33473cd80da4cc05d5ab2c46b0a7

File: op-e2e/actions/sync_test.go (hunk 11 above)
Comment (mininny) [reply op-e2e/actions/sync_test.go:790]: I deduplicated the tests here :+1: : https://github.com/ethereum-optimism/optimism/pull/10767/commits/5114a0ad2f5385a387fa980f3de44dfd998e651b

File: op-node/rollup/driver/driver.go (hunk 12 above)
Comment (mininny) [reply op-node/rollup/driver/driver.go]: Thanks! I agree with you, and it makes sense since `SupportsPostFinalizationELSync` indeed is a configuration for the syncing behavior. I updated the code to use the engine kind from the flags and feed it directly into the `sync.Config`: https://github.com/ethereum-optimism/optimism/pull/10767/commits/ac208ef3fef13a3340ec0391cbb17628dae420b4

**Comment by joshieDo** [comment]
An alternative to a configuration flag could be using [engine_getClientVersionV1](https://github.com/ethereum/execution-apis/blob/main/src/engine/identification.md#engine_getclientversionv1) to get the client, and choose the sync mode from there

**Comment by axelKingsley** [comment]
> An alternative to a configuration flag could be using [engine_getClientVersionV1](https://github.com/ethereum/execution-apis/blob/main/src/engine/identification.md#engine_getclientversionv1) to get the client, and choose the sync mode from there

I like this idea, but to avoid scope-creep this can be taken iteratively. For example, in the future, the flag could be set to `autoDetect` to run the `engine_getClientVersionV1` determination path, but if set manually we can use that instead. And for now, to distinguish the sync behavior, just making distinct geth/reth is a good start.

**Review by protolambda — APPROVED** [review]
